### PR TITLE
Delete recipe

### LINKED ISSUE #8: As a user, I want to be able to delete a recipe that is no longer relevant, so that I can keep my list updated and clean.
D for recipe class.

### PR DESCRIPTION
**Pull request by khenghun**
Enable deletion of recipe

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/DeleteRecipeCommand.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    public static final String COMMAND_WORD = "delr";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Deletes a recipe in the recipe list. "
+            + "Parameters: "
+            + PREFIX_RECIPE_PRODUCT_NAME + "PRODUCT NAME "
+            + PREFIX_RECIPE_ID + "RECIPE ID ";
+
+    public static final String MESSAGE_SUCCESS = "Recipe has been deleted: %1$s";
+    public static final String MESSAGE_RECIPE_NOT_FOUND = "Recipe is not found in the recipe list";
+    public static final String MESSAGE_INDEX_NOT_FOUND = "Id is out of range";
```
Comment (aqcd) [inline src/main/java/seedu/address/logic/commands/DeleteRecipeCommand.java]: for clarity can change to "Recipe ID is out of range"

File: src/main/java/seedu/address/logic/commands/DeleteRecipeCommand.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+
+    private final String productName;
+    private final Index index;
+
+    /**
+     * Creates an DeleteRecipeCommand to delete the specified {@code Item}
```
Comment (aqcd) [inline src/main/java/seedu/address/logic/commands/DeleteRecipeCommand.java]: should this be {@code Item}?

File: src/main/java/seedu/address/logic/commands/DeleteRecipeCommand.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        List<Recipe> recipelist = new ArrayList<>(model.getFilteredRecipeList());
+        // filter to only get matching and not deleted recipes
+        recipelist.removeIf(x -> !x.getProductName().equals(productName) || x.isDeleted());
+        if (recipelist.isEmpty()) {
+            throw new CommandException(MESSAGE_RECIPE_NOT_FOUND);
+        }
+        Recipe recipeToDelete;
+        try {
+            recipeToDelete = recipelist.get(index.getOneBased() - 1);
+        } catch (Exception e) {
```
Comment (aqcd) [inline src/main/java/seedu/address/logic/commands/DeleteRecipeCommand.java]: catch specific exception

File: src/main/java/seedu/address/logic/parser/CliSyntax.java [hunk 4]
```diff
@@ -23,5 +23,6 @@
     public static final Prefix PREFIX_RECIPE_INGREDIENTS = new Prefix("-items ");
     public static final Prefix PREFIX_RECIPE_PRODUCT_QUANTITY = new Prefix("-pc ");
     public static final Prefix PREFIX_RECIPE_DESCRIPTION = new Prefix("-d ");
+    public static final Prefix PREFIX_RECIPE_ID = new Prefix("-i ");
```
Comment (aqcd) [inline src/main/java/seedu/address/logic/parser/CliSyntax.java:27]: add comment above this line specifying that this is for the delr command

File: src/main/java/seedu/address/logic/parser/DeleteRecipeCommandParser.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+/**
+ * Parses input arguments and creates a new DeleteRecipeCommand object
+ */
+public class DeleteRecipeCommandParser implements Parser<DeleteRecipeCommand> {
+
+    /**
+     * Parses the given {@code String} of arguments in the context of the DeleteRecipeCommand
+     * and returns a DeleteRecipeCommand object for execution.
+     *
+     * @throws ParseException if the user input does not conform the expected format
```
Comment (aqcd) [inline src/main/java/seedu/address/logic/parser/DeleteRecipeCommandParser.java]: conform -> conform to

File: src/main/java/seedu/address/model/RecipeList.java [hunk 6]
```diff
@@ -117,4 +125,6 @@ public boolean equals(Object other) {
     public int hashCode() {
         return recipes.hashCode();
     }
+
+
```
Comment (aqcd) [inline src/main/java/seedu/address/model/RecipeList.java]: extra lines

File: src/main/java/seedu/address/model/recipe/Recipe.java [hunk 7]
```diff
@@ -38,6 +38,20 @@ public Recipe(int id, IngredientList ingredients, int productId, String productN
...
+        ProductQuantity productQuantity, String description) {
+        this.id = id;
+        this.ingredients = ingredients;
+        this.productId = productId;
+        this.productName = productName;
+        this.productQuantity = productQuantity;
+        this.description = description;
+        this.isDeleted = true;
+    }
+
```
Comment (aqcd) [inline src/main/java/seedu/address/model/recipe/Recipe.java]: this constructor might be misleading, would it be better to modify the existing one?
for instance
```
if (!isDeleted) {
    idCounter++;
}
```
in the existing constructor might make it usable as an equivalent to this one

File: src/main/java/seedu/address/model/recipe/UniqueRecipeList.java [hunk 8]
```diff
@@ -79,6 +79,21 @@ public void remove(Recipe toRemove) {
         }
     }
 
+    /**
+     * Deletes the equivalent recipe from the list, sets deleted flag to true.
+     * The recipe must exist in the list.
+     */
+    public void delete(Recipe toDelete) {
+        requireNonNull(toDelete);
+        System.out.println(1);
```
Comment (aqcd) [inline src/main/java/seedu/address/model/recipe/UniqueRecipeList.java]: remove or change to logger statement

**Review by aqcd — COMMENTED** [review]
Preliminary review.

File: src/main/java/seedu/address/logic/commands/DeleteRecipeCommand.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        requireNonNull(model);
+        List<Recipe> recipelist = new ArrayList<>(model.getFilteredRecipeList());
+        // filter to only get matching and not deleted recipes
+        recipelist.removeIf(x -> !x.getProductName().equals(productName) || x.isDeleted());
+        if (recipelist.isEmpty()) {
+            throw new CommandException(MESSAGE_RECIPE_NOT_FOUND);
+        }
+        Recipe recipeToDelete;
+        try {
+            recipeToDelete = recipelist.get(index.getOneBased() - 1);
```
Comment (justacasul) [inline src/main/java/seedu/address/logic/commands/DeleteRecipeCommand.java]: index.getOneBased() - 1 is just index.getZeroBased()?

File: src/main/java/seedu/address/logic/parser/RecipeParserUtil.java [hunk 10]
```diff
@@ -78,4 +78,5 @@ public static String parseDescription(String description) {
         requireNonNull(description);
         return description.trim();
     }
+
```
Comment (justacasul) [inline src/main/java/seedu/address/logic/parser/RecipeParserUtil.java:81]: what is this haha

File: src/main/java/seedu/address/model/recipe/UniqueRecipeList.java [hunk 11]
```diff
@@ -79,6 +79,21 @@ public void remove(Recipe toRemove) {
...
+        if (!internalList.contains(toDelete)) {
+            throw new RecipeNotFoundException();
+        }
+        Recipe toAdd = toDelete.delete();
```
Comment (justacasul) [inline src/main/java/seedu/address/model/recipe/UniqueRecipeList.java]: toAdd is can get confusing here, consider another name (deletedRecipe/updatedRecipe/etc)?

File: src/main/java/seedu/address/model/recipe/UniqueRecipeList.java (hunk 8 above)
Comment (khenghun) [reply src/main/java/seedu/address/model/recipe/UniqueRecipeList.java]: oops forgot to remove this

**Review by justacasul — COMMENTED** [review]
Seems good

**Pull request merged**